app/ModelFunctions/LDAPFunctions.php

```diff
@@ -0,0 +1,377 @@
+			if (!@ldap_bind($this->con, $this->getConf('binddn'), $this->getConf('bindpw'))) {
+				Logs::notice(__METHOD__, __LINE__, 'LDAP bind as superuser failed.');
+
+				return false;
+			}
+			$this->bound = 2;
+		}
+
+		$info = [];
+		$info['user'] = $user;
+		$info['server'] = $this->getConf('server');
+
+		// get info for given user
+		$base = $this->_makeFilter($this->getConf('usertree'), $info);
+		$this->_debug(__METHOD__, __LINE__, 'base filter: ' . $base);
+		if ($this->getConf('userfilter')) {
+			$filter = $this->_makeFilter($this->getConf('userfilter'), $info);
+		} else {
+			$filter = '(ObjectClass=*)';
+		}
+		$this->_debug(__METHOD__, __LINE__, 'filter: ' . $filter);
+
+		$sr = $this->_ldapsearch($this->con, $base, $filter, $this->getConf('userscope'));
+		$result = @ldap_get_entries($this->con, $sr);
+		ldap_free_result($sr);
+
+		// if result is not an array
+		if (!is_array($result)) {
+			// no objects found
+			Logs::notice(__METHOD__, __LINE__, 'LDAP search returned non-array result.');
+
+			return false;
+		}
+
+		// Only accept one response
+		if ($result['count'] != 1) {
+			Logs::notice(__METHOD__, __LINE__, 'LDAP search returned ' . $result['count'] . ' results while it should return 1!');
+
+			return false;
+		}
+
+		$user_result = $result[0];
+
+		// general user info
+		$info['dn'] = $user_result['dn'];
+		$info['fullname'] = $user_result[$this->getConf('cn', 'cn')][0];
+		if (array_key_exists($this->getConf('mail', 'mail'), $user_result)) {
+			$info['email'] = $user_result[$this->getConf('mail', 'mail')][0];
+		} else {
+			$info['email'] = '';
+		}
+		// cache the info for future use
+		$this->user_info[$user] = $info;
+
+		return $info;
+	}
+
+	/**
+	 * Check user+password.
+	 *
+	 * Checks if the given user exists and the given
+	 * plaintext password is correct by trying to bind
+	 * to the LDAP server
+	 *
+	 * @param string $user
+	 * @param string $pass
+	 *
+	 * @return bool
+	 */
+	public function check_pass($user, $pass)
+	{
+		if (!$this->_openLDAP()) {
+			return false;
+		}
+
+		// indirect user bind
+		if ($this->getConf('binddn') && $this->getConf('bindpw')) {
+			// use superuser credentials
+			if (!@ldap_bind($this->con, $this->getConf('binddn'), $this->getConf('bindpw'))) {
+				Logs::notice(__METHOD__, __LINE__, 'LDAP bind as superuser failed.');
+
+				return false;
+			}
+			$this->bound = 2;
+		} elseif ($this->getConf('binddn') &&
+			$this->getConf('usertree') &&
+			$this->getConf('userfilter')
+		) {
+			// special bind string
+			$dn = $this->_makeFilter(
+				$this->getConf('binddn'),
+				['user' => $user, 'server' => $this->getConf('server')]
+			);
+		} elseif (strpos($this->getConf('usertree'), '%{user}')) {
+			// direct user bind
+			$dn = $this->_makeFilter(
+				$this->getConf('usertree'),
+				['user' => $user, 'server' => $this->getConf('server')]
+			);
+		} else {
+			// Anonymous bind
+			if (!@ldap_bind($this->con, $this->getConf('binddn'), $this->getConf('bindpw'))) {
```

Comment:
Please don't use magic string constants. Here, I am specifically referring to `'binddn'`  and `'bindpw'`. Introduce proper class constants, e.g. `const CONF_BIND_DN = 'ldap_binddn';` at the top of the class and then use those constants.